dev/tests/integration/testsuite/Magento/Customer/Model/ResourceModel/AddressRepositoryTest.php

```diff
@@ -327,8 +327,11 @@ public function testSearchAddresses($filters, $filterGroup, $filterOrders, $expe
             }
         }
 
-        $searchBuilder->setPageSize(1);
-        $searchBuilder->setCurrentPage(2);
+        $pageSize = 1;
+        $expectedPageCount = ceil(count($expectedResult) / $pageSize);
```

Comment:
This is not a good idea to put logic into test. Why this integration is even affected by API-only changes?

Reply:
`$searchBuilder = $this->_objectManager->create(\Magento\Framework\Api\SearchCriteriaBuilder::class);` ah, ok, so this seems to be widely used in the system besides API.

Reply:
Yes. The API's code is used outside of the main API.

I've also found, while diagnosing the failed tests, that two assumptions exist in several places:

1. **Page `< 1` should retrieve Page 1.** I could not find the source of this assumption. I set the API to assume that any page number less than 1 should be set to Page 1 to avoid breaking existing behavior. We had previously discussed underflow, and this abides by that discussion.
2. **Page `max + 1` should return Page `max`**. In this instance, the test was passing because Page 2 of 1 was returning Page 1. In the AddressRepositoryTest, there are tests that should return only 1 page, but the test always sets "Page 2". A side-effect of the behavior this PR addresses is that the tests *could* pass if only 1 address were returned (because Page 2 would return Page 1).

I haven't fixed all of the failures from the integration tests, yet, but all of the issues I've found recently are because of the two points I listed above.

Reply:
```
$pageSize = 1;
$expectedPageCount = ceil(count($expectedResult) / $pageSize);
```

is it different from 
```
$expectedPageCount = count($expectedResult);
```
?

1) There is no place between WebAPI and search builder which could be changed to not affect integration tests at all but only WebAPI?
2) If so, maybe it makes more sense to affect Web UI pagination as well - by changing underlying collection logic and nothing more?

Reply:
> {code} is it different from {code}
They are effectively the same. The first one is just for clarity, but could be easily updated to the second one. Would you prefer the 1-line version over the 2-line version?

> 1. There is no place between the WebAPI and search builder which could be changed to not affect integration tests at all but only WebAPI?

These changes invalidate the assumption that `(2 of 1) === (1 of 1)`. Any tests that make this assumption at the API level (such as this one) will need to be updated regardless of where the fix is applied.

> 2. If so, maybe it makes more sense to affect Web UI pagination as well - by changing underlying collection logic and nothing more?

This is an option, but it would affect more of the system.  This change would be at `./lib/internal/Magento/Framework/Data/Collection.php:239`:

```
    public function getCurPage($displacement = 0)
    {
        if ($this->_curPage + $displacement < 1) {
            return 1;
/* These lines would be removed
        } elseif ($this->_curPage + $displacement > $this->getLastPageNumber()) {
            return $this->getLastPageNumber();
*/
        } else {
            return $this->_curPage + $displacement;
        }
    }
```
However, tests that assume that Page 2 of 1 should return Page 1 would still need to be updated. That's the logic being changed.

Reply:
Also, tests like this would need to go:

```php
    public function testGetCurPage()
    {
        $this->_model->setCurPage(10);
        $this->assertEquals(1, $this->_model->getCurPage());
    }
```

Reply:
Also, this test data would need to change:

``` php
    public function getPriceFormatDataProvider()
    {
        return [
            ['en_US', ['decimalSymbol' => '.', 'groupSymbol' => ',']],
            ['de_DE', ['decimalSymbol' => ',', 'groupSymbol' => '.']],
            ['de_CH', ['decimalSymbol' => '.', 'groupSymbol' => '\'']], //<--- Right Here
            ['uk_UA', ['decimalSymbol' => ',', 'groupSymbol' => ' ']]
        ];
    }
```

That value should be `’`, not `\'`. This test should not be passing right now.